beacon-chain/blockchain/forkchoice/service.go

```diff
@@ -0,0 +1,460 @@
+	head := s.justifiedCheckpt.Root
+
+	for {
+		startSlot := s.justifiedCheckpt.Epoch * params.BeaconConfig().SlotsPerEpoch
+		filter := filters.NewFilter().SetParentRoot(head).SetStartSlot(startSlot)
+		children, err := s.db.BlockRoots(s.ctx, filter)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not retrieve children info")
+		}
+
+		if len(children) == 0 {
+			return head, nil
+		}
+
+		// if a block has one child, then we don't have to lookup anything to
+		// know that this child will be the best child.
+		head = children[0]
+		if len(children) > 1 {
+			highest, err := s.latestAttestingBalance(head)
+			if err != nil {
+				return nil, errors.Wrap(err, "could not get latest balance")
+			}
+			for _, child := range children[1:] {
+				balance, err := s.latestAttestingBalance(child)
+				if err != nil {
+					return nil, errors.Wrap(err, "could not get latest balance")
+				}
+
+				if balance > highest {
+					highest = balance
+					head = child
+				}
+			}
+		}
+	}
+}
+
+// OnTick tracks the last unix time of when a block or an attestation arrives.
+//
+// Spec pseudocode definition:
+//   def on_tick(store: Store, time: uint64) -> None:
+//    store.time = time
+func (s *Store) OnTick(t uint64) {
+	s.time = t
+}
+
+// OnBlock is called whenever a block is received. It runs state transition on the block and
+// update fork choice store struct.
+//
+// Spec pseudocode definition:
+//   def on_block(store: Store, block: BeaconBlock) -> None:
+//    # Make a copy of the state to avoid mutability issues
+//    assert block.parent_root in store.block_states
+//    pre_state = store.block_states[block.parent_root].copy()
+//    # Blocks cannot be in the future. If they are, their consideration must be delayed until the are in the past.
+//    assert store.time >= pre_state.genesis_time + block.slot * SECONDS_PER_SLOT
+//    # Add new block to the store
+//    store.blocks[signing_root(block)] = block
+//    # Check block is a descendant of the finalized block
+//    assert (
+//        get_ancestor(store, signing_root(block), store.blocks[store.finalized_checkpoint.root].slot) ==
+//        store.finalized_checkpoint.root
+//    )
+//    # Check that block is later than the finalized epoch slot
+//    assert block.slot > compute_start_slot_of_epoch(store.finalized_checkpoint.epoch)
+//    # Check the block is valid and compute the post-state
+//    state = state_transition(pre_state, block)
+//    # Add new state for this block to the store
+//    store.block_states[signing_root(block)] = state
+//
+//    # Update justified checkpoint
+//    if state.current_justified_checkpoint.epoch > store.justified_checkpoint.epoch:
+//        store.justified_checkpoint = state.current_justified_checkpoint
+//
+//    # Update finalized checkpoint
+//    if state.finalized_checkpoint.epoch > store.finalized_checkpoint.epoch:
+//        store.finalized_checkpoint = state.finalized_checkpoint
+func (s *Store) OnBlock(b *ethpb.BeaconBlock) error {
```

Comment:
This function does a lot of stuff, any chance it can be broken up into smaller funcs with clear naming?

Reply:
yep